onnxruntime/core/optimizer/common_subexpression_elimination.cc

```diff
@@ -0,0 +1,428 @@
+    while (!arg.empty() && arg.back()->output_index_ == kInvalidOutputIndex && arg.back()->non_op_value_ == nullptr) {
+      arg.pop_back();
+    }
+  }
+
+  return result;
+}
+
+template <typename Range>
+bool AreRangesEqual(const Range& lhs, const Range& rhs) {
+  return std::equal(lhs.begin(), lhs.end(), rhs.begin(), rhs.end());
+}
+
+bool AreEqual(const ONNX_NAMESPACE::AttributeProto& lhs, const ONNX_NAMESPACE::AttributeProto& rhs) {
+  if (&lhs == &rhs) {
+    return true;
+  }
+
+  if (lhs.type() != rhs.type() || lhs.name() != rhs.name()) {
+    return false;
+  }
+
+  switch (lhs.type()) {
+    case onnx::AttributeProto_AttributeType_FLOAT:
+      return lhs.f() == rhs.f();
+    case onnx::AttributeProto_AttributeType_INT:
+      return lhs.i() == rhs.i();
+    case onnx::AttributeProto_AttributeType_STRING:
+      return lhs.s() == rhs.s();
+    case onnx::AttributeProto_AttributeType_FLOATS:
+      return AreRangesEqual(lhs.floats(), rhs.floats());
+    case onnx::AttributeProto_AttributeType_INTS:
+      return AreRangesEqual(lhs.ints(), rhs.ints());
+    case onnx::AttributeProto_AttributeType_STRINGS:
+      return AreRangesEqual(lhs.strings(), rhs.strings());
+    case onnx::AttributeProto_AttributeType_TENSOR:
+    case onnx::AttributeProto_AttributeType_GRAPH:
+    case onnx::AttributeProto_AttributeType_SPARSE_TENSOR:
+    case onnx::AttributeProto_AttributeType_TENSORS:
+    case onnx::AttributeProto_AttributeType_GRAPHS:
+    case onnx::AttributeProto_AttributeType_SPARSE_TENSORS:
+    case onnx::AttributeProto_AttributeType_UNDEFINED:
+      return false; // Don't support these attributes for now; corresponding nodes will be considered distinct.
+  }
+
+  return false;
+}
+
+std::size_t GetAttributeHash(const ONNX_NAMESPACE::AttributeProto& attr) {
+  std::size_t hash = 0;
+  UpdateHash(attr.type(), hash);
+  UpdateHash(attr.name(), hash);
+  switch (attr.type()) {
+    case onnx::AttributeProto_AttributeType_FLOAT:
+      UpdateHash(attr.f(), hash);
+      break;
+    case onnx::AttributeProto_AttributeType_INT:
+      UpdateHash(attr.i(), hash);
+      break;
+    case onnx::AttributeProto_AttributeType_STRING:
+      UpdateHash(attr.s(), hash);
+      break;
+    case onnx::AttributeProto_AttributeType_FLOATS:
+      UpdateHashWithContainer(attr.floats(), hash);
+      break;
+    case onnx::AttributeProto_AttributeType_INTS:
+      UpdateHashWithContainer(attr.ints(), hash);
+      break;
+    case onnx::AttributeProto_AttributeType_STRINGS:
+      UpdateHashWithContainer(attr.strings(), hash);
+      break;
+    case onnx::AttributeProto_AttributeType_TENSOR:
+    case onnx::AttributeProto_AttributeType_GRAPH:
+    case onnx::AttributeProto_AttributeType_SPARSE_TENSOR:
+    case onnx::AttributeProto_AttributeType_TENSORS:
+    case onnx::AttributeProto_AttributeType_GRAPHS:
+    case onnx::AttributeProto_AttributeType_SPARSE_TENSORS:
+    case onnx::AttributeProto_AttributeType_UNDEFINED:
+      break;
+  }
```

Comment:
Should this just break or error out?

Reply:
Same hash for different values is expected (consider that there is a finite number of hashes but infinite number of attributes). The hash map will do collision resolution.